core/src/main/java/org/apache/calcite/rel/rules/FilterExtractInnerJoinRule.java

```diff
@@ -0,0 +1,194 @@
+
+    call.transformTo(modifiedJoinClauseWithWhereClause);
+  }
+
+  private static boolean isCrossJoin(Join join, RelBuilder builder) {
+    return join.getJoinType().equals(JoinRelType.INNER)
+        && builder.literal(true).equals(join.getCondition());
+  }
+  private void populateStackWithEndIndexesForTables(
+      Join join, Stack<Map.Entry<RelNode, Integer>> stack, List<RexNode> joinConditions) {
+    RelNode left = ((HepRelVertex) join.getLeft()).getCurrentRel();
+    RelNode right = ((HepRelVertex) join.getRight()).getCurrentRel();
+    int leftTableColumnSize = join.getLeft().getRowType().getFieldCount();
+    int rightTableColumnSize = join.getRight().getRowType().getFieldCount();
+    stack.push(Map.entry(right, leftTableColumnSize + rightTableColumnSize - 1));
+    if (!(join.getCondition() instanceof RexLiteral)) {
+      joinConditions.add(join.getCondition());
+    }
+    if (left instanceof Join) {
+      populateStackWithEndIndexesForTables((Join) left, stack, joinConditions);
+    } else {
+      stack.push(Map.entry(left, leftTableColumnSize - 1));
+    }
+  }
+
+  private RelNode moveConditionsFromWhereClauseToJoinOnClause(
+      List<RexNode> allConditions, Stack<Map.Entry<RelNode, Integer>> stack, RelBuilder builder) {
+    Map.Entry<RelNode, Integer> leftEntry = stack.pop();
+    Map.Entry<RelNode, Integer> rightEntry;
+    RelNode left = leftEntry.getKey();
+
+    while (!stack.isEmpty()) {
+      rightEntry = stack.pop();
+      List<RexNode> joinConditions = getConditionsForEndIndex(allConditions, rightEntry.getValue());
+      RexNode joinPredicate;
+      if (joinConditions.size() == 2
+          && joinConditions.stream().allMatch(condition -> condition instanceof RexInputRef)) {
+        joinPredicate = builder.equals(joinConditions.get(0), joinConditions.get(1));
+      } else {
+        joinPredicate = builder.and(joinConditions);
+      }
+      allConditions.removeAll(joinConditions);
+      left = LogicalJoin.create(left, rightEntry.getKey(), ImmutableList.of(),
+          joinPredicate, ImmutableSet.of(), JoinRelType.INNER);
+    }
+    return builder.push(left)
+        .filter(builder.and(allConditions))
+        .build();
+  }
+
+  private List<RexNode> getConditionsForEndIndex(List<RexNode> conditions, int endIndex) {
+    if (conditions.stream().allMatch(condition -> condition instanceof RexInputRef
+        && ((RexInputRef) condition).getIndex() <= endIndex)) {
+      return conditions;
+    }
+    return conditions.stream()
+        .filter(
+            condition ->
+                ((RexCall) condition).operands.stream().noneMatch(
+                    operand -> operand instanceof RexLiteral)
+                    && ((RexCall) condition).operands.size() > 1
+                    && checkEndIndex((RexCall) condition, endIndex)
+        )
+        .collect(Collectors.toList());
+  }
+
+  private boolean checkEndIndex(RexCall condition, int endIndex) {
```

Comment:
Change the name that starts with "is" since it returns a boolean.

Reply:
Done!